tests/test_request/test_request.py

```diff
@@ -0,0 +1,276 @@
+        req.submit()
+        assert work.done.wait(timeout=1)
+
+        req.notify_finished(recv)
+
+        recv.assert_called_once_with(42)
+
+    def test_signal_finished_should_not_be_called_on_exception(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+        recv = mock.Mock()
+
+        req.notify_finished(recv)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+        recv.assert_not_called()
+
+    def test_signal_failed_should_be_called_on_exception(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+
+        recv = mock.Mock()
+
+        req = Request(work)
+        req.notify_failed(recv)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+        recv.assert_called_once()
+        assert isinstance(recv.call_args[0][0], TExc)
+
+    def test_signal_failed_called_even_when_subscription_happened_after_completion(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+
+        recv = mock.Mock()
+
+        req = Request(work)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+
+        req.notify_failed(recv)
+        recv.assert_called_once()
+        assert isinstance(recv.call_args[0][0], TExc)
+
+
+@pytest.fixture
+def work():
+    unpause = threading.Event()
+    children = []
+
+    def work_fn():
+        more_work = Request(lambda: 42)
+        some_more_work = Request(lambda: 42)
+        children.extend([more_work, some_more_work])
+        unpause.wait()
+        return more_work.wait()
+
+    work = Work(work_fn)
+    work.unpause = unpause
+
+    work.request = Request(work)
+    work.request.submit()
+    work.children = children
+
+    assert work.started.wait()
+
+    yield work
+
+    if not work.unpause.is_set():
+        work.unpause.set()
+
+    assert work.done.wait()
+
+
+def test_requests_created_within_request_considired_child_requests(work):
+    assert work.request.child_requests == set(work.children)
+    assert len(work.request.child_requests) == 2
+
+
+@pytest.fixture
+def cancelled_work(work):
+    assert not work.request.cancelled
+    work.request.cancel()
+    assert work.request.cancelled
+
+    work.unpause.set()
+    assert work.done.wait()
+
+    return work
+
+
+def test_wait_for_cancelled_rq_raises_invalid_request_exception(cancelled_work):
+    work_rq = cancelled_work.request
+
+    with pytest.raises(Request.InvalidRequestException):
+        assert work_rq.wait()
+
+
+def test_cancel_raises_exception_on_yield_point(cancelled_work):
+    work_rq = cancelled_work.request
+    assert isinstance(cancelled_work.exception, Request.CancellationException)
+
+
+def test_cancels_child_requests(cancelled_work):
+    assert len(cancelled_work.children) == 2
+
+    for ch in cancelled_work.children:
+        assert ch.cancelled
+
+    work_rq = cancelled_work.request
+
+    assert work_rq.child_requests == set()
```

Comment:
```suggestion
    assert not work_rq.child_requests
```

Unless `child_requests` could be something else than `set`.